### PR TITLE
[SPARK-34033][R] SparkR Daemon Initialization

### PR DESCRIPTION
**Pull request by WamBamBoozle**
I closed this because I broke the git history. Please see #31162

### What changes were proposed in this pull request?

A way of letting sparkR users preload libraries and other initialization
in the daemon before it starts forking. A full description can be found
in docs/sparkr.md.

### Why are the changes needed?

Performance

### Does this PR introduce _any_ user-facing change?

Yes. It introduces two new sparkR session parameters,
spark.r.daemonInit and spark.r.daemonTimeout, as described in
docs/configuration.md.

### How was this patch tested?

added pkg/tests/fulltests/test_daemon_initialization.R

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/internal/config/R.scala [hunk 1]
```diff
@@ -42,4 +42,16 @@ private[spark] object R {
...
+  val R_DAEMON_INIT = ConfigBuilder("spark.r.daemonInit")
+    .version("3.2.0")
+    .stringConf
+    .createWithDefault("NULL")
+
+  /** milliseconds to wait until we give up launching the daemon */
+  val R_DAEMON_TIMEOUT = ConfigBuilder("spark.r.daemonTimeout")
+    .version("3.2.0")
+    .intConf
+    .createWithDefault(10000)
```
Comment (WamBamBoozle) [inline core/src/main/scala/org/apache/spark/internal/config/R.scala:56]: Is the version indicated here reasonable? I don't understand the criteria for assigning version numbers really

**Review by HyukjinKwon — COMMENTED** [review]
@WamBamBoozle, can we have something like `spark.python.daemon.module` and `spark.python.worker.module` to allow custom worker and daemon instead?

**Comment by WamBamBoozle** [comment]
> can we have something like spark.python.daemon.module and spark.python.worker.module to allow custom worker and daemon instead?

No. Those configurations put a lot of responsibility on the user. Further, it is very prone to breakage due to changes in Spark. 

What this PR proposes is a narrow thing: initializing the daemon. It doesn't change the logic of either the daemon or the worker.

**Comment by HyukjinKwon** [comment]
@WamBamBoozle, both configurations `spark.python.daemon.module` and `spark.python.worker.module` came out to cover a lot of stories. For example, do something like post-hoc, initialisation, wrapping the computation, etc. I would definitely like to avoid adding each configuration for each case.

**Comment by WamBamBoozle** [comment]
Greetings @HyukjinKwon

Thanks so much for your review!

I'm looking at [python/test_coverage/coverage_daemon.py](https://github.com/apache/spark/pull/20204/files#diff-a19676926d6ad7cb2306c47aedc8c9dfd5bc4413ce1aaae8fb3b04c7c5836110)

I agree that spark.python.daemon.module is more general and more powerful than what I am proposing. This PR is for fixing one thing, whereas what you did was a very general solution to a different problem. I don't want to warn my users that this is for advanced users only. The concept is simple: put your initialization code here.

I propose we provide

- spark.r.daemon.finalize
- spark.r.daemon.initialize
- spark.r.worker.finalize
- spark.r.worker.initialize

where the value of each of those are strings of R code

that way we've achieved the same ends you have without forcing the user to understand the minutia of how and what to load. If you agree, I will change this PR to be the implementation of spark.r.daemon.initialize (instead of what it is now, spark.r.daemonInit). I would then open another ticket to implement the other flags.

**Pull request closed without merging**